internal/labels/helpers.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package labels contains functions to validate and compare values used in Kubernetes labels.
+package labels
+
+import (
+	"encoding/base64"
+	"hash/fnv"
+
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+// MustFormatValue returns the passed inputLabelValue if it meets the standards for a Kubernetes label value.
+// If the name is not a valid label value this function returns a hash which meets the requirements.
+func MustFormatValue(str string) string {
+	// a valid Kubernetes label value must:
+	// - be less than 64 characters long.
+	// - be an empty string OR consist of alphanumeric characters, '-', '_' or '.'.
+	// - start and end with an alphanumeric character
+	if len(validation.IsValidLabelValue(str)) == 0 {
+		return str
+	}
+	hasher := fnv.New32a()
+	_, err := hasher.Write([]byte(str))
+	if err != nil {
+		// At time of writing the implementation of fnv's Write function can never return an error.
+		// If this changes in a future go version this function will panic.
+		panic(err)
+	}
+	return base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(hasher.Sum(nil))
+}
+
+// MustEqualValue returns true if the actualLabelValue equals either the inputLabelValue or the hashed
+// value of the inputLabelValue.
+func MustEqualValue(str, labelValue string) bool {
+	expectedLabelValue := MustFormatValue(str)
+	if labelValue == expectedLabelValue {
+		return true
+	}
+	hasher := fnv.New32a()
+	// The Write function should never return an error.
+	_, err := hasher.Write([]byte(labelValue))
+	if err != nil {
+		// At time of writing the implementation of fnv's Write function can never return an error.
+		// If this changes in a future go version this function will panic.
+		panic(err)
+	}
+	return expectedLabelValue == base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(hasher.Sum(nil))
```

Comment:
We are doing this instead of:
```go
return MustFormatValue(str) == MustFormatValue(labelValue)

```
because we want to compare hashed `str` value with hash `labelValue` even if `labelValue` is less than 64 characters, isn't it?
When will we hit such a case?

Reply:
We could do `MustFomatValue` as the hash will always be shorter than the 63 characters meaning the output value will be the same as the input value, but I think it's clearer and easier to reason about if we step through the comparison cases here. I'm happy to change if there's strong opinions to the contrary though.

Reply:
I am fine with keeping the current function block as is if we adjust all the uses of this function to be consistent with what we pass as `str` and `labelValue`. In some cases we are using the target object name as `labelValue` and in some cases we are using the target label value as `labelValue`.
- https://github.com/ykakarap/cluster-api/blob/3fea53de24307d776164c49342f5fdfcb081e508/controlplane/kubeadm/internal/controllers/controller.go#L346
- https://github.com/ykakarap/cluster-api/blob/3fea53de24307d776164c49342f5fdfcb081e508/internal/controllers/machineset/machineset_controller.go#L303

Reply:
Updated for that bug and simplified the function. You were right that the previous version was over-covering the cases. I think the simplified version of comparing a formatted version of the string with the inputted labelValue is enough to cover all cases.